### PR TITLE
Unit tests workspace part1

### PR DESCRIPTION
**Pull request by DoctorHeisenberg**
Unit tests for Workspace service

### CODE REVIEW

**Review by AndriyKovalets — APPROVED** [review]

File: Provis.UnitTests/Base/TestData/InviteUserTestData.cs [hunk 1]
```diff
@@ -36,5 +37,36 @@ public static List<InviteUser> GetInviteUserList()
...
+                    ToUserId = "1",
+                    IsConfirm = false,
+                    WorkspaceName = "worksace2"
+                },
+                new UserInviteInfoDTO()
+                {
+                    Id = 1,
+                    FromUserName = "username2",
+                    ToUserId = "1",
+                    IsConfirm = true,
```
Comment (YevheniiVovk) [inline Provis.UnitTests/Base/TestData/InviteUserTestData.cs]: For a what reason this class exist ?

File: Provis.UnitTests/Base/TestData/UserTaskTestData.cs [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+                        TaskId = 1,
+                        UserId = "1",
+                        UserRoleTagId = 1,
+                        IsUserDeleted = false
+                    },
+                    3,4,"Test1")
+            };
+        }
+    }
+}
```
Comment (YevheniiVovk) [inline Provis.UnitTests/Base/TestData/UserTaskTestData.cs]: Remove this static methods please and another one class too. 
Create a with methods in scope of test class

File: Provis.UnitTests/Base/TestData/UserWorkspaceTestData.cs [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+                        Name = "Name4",
+                        Surname = "Surname4",
+                        UserName = "Username4",
+                        ImageAvatarUrl = "Path4"
+                    }
+                }
+            };
+        }
+    }
+}
```
Comment (YevheniiVovk) [inline Provis.UnitTests/Base/TestData/UserWorkspaceTestData.cs]: Same

File: Provis.UnitTests/Base/TestData/WorkspaceTestData.cs [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+        {
+            return new WorkspaceChangeRoleDTO()
+            {
+                UserId = "4",
+                RoleId = 3,
+                WorkspaceId = 1
+            };
+        }
+    }
+}
```
Comment (YevheniiVovk) [inline Provis.UnitTests/Base/TestData/WorkspaceTestData.cs]: Same

File: Provis.UnitTests/Core/Services/WorkspaceServiceTests.cs [hunk 5]
```diff
@@ -87,5 +104,522 @@ public void TearDown()
             _optionsMock.Verify();
             _metricsMock.Verify();
         }
+
+        [Test]
+        [TestCase("1")]
```
Comment (YevheniiVovk) [inline Provis.UnitTests/Core/Services/WorkspaceServiceTests.cs]: For a what reason you use testcase here ?

File: Provis.UnitTests/Core/Services/WorkspaceServiceTests.cs [hunk 6]
```diff
@@ -87,5 +104,522 @@ public void TearDown()
...
+        public async Task CreateWorkspaceAsync_UserIsValidAndDTOIsValid_ReturnCompletedTask(string userid)
+        {
+            var workspaceCreateDTOMock = WorkspaceTestData.GetWorkspaceCreateDTO();
+            var workspaceMock = WorkspaceTestData.GetTestWorkspace();
```
Comment (YevheniiVovk) [inline Provis.UnitTests/Core/Services/WorkspaceServiceTests.cs]: It must be changed

File: Provis.UnitTests/Core/Services/WorkspaceServiceTests.cs [hunk 7]
```diff
@@ -87,5 +104,522 @@ public void TearDown()
...
+
+            SetupAddWorkspaceAsync(workspaceMock);
+            SetupWorkspaceSaveChangesAsync();
+
+            var userWorkspaceMock = UserWorkspaceTestData.GetTestUserWorkspaceList()[0];
```
Comment (YevheniiVovk) [inline Provis.UnitTests/Core/Services/WorkspaceServiceTests.cs]: What a [0] ?

File: Provis.UnitTests/Core/Services/WorkspaceServiceTests.cs [hunk 8]
```diff
@@ -87,5 +104,522 @@ public void TearDown()
...
+
+            SetupAddUserWorkspaceAsync(userWorkspaceMock);
+            SetupUserWorkspaceSaveChangesAsync();
+
+            var result = _workspaceService.CreateWorkspaceAsync(workspaceCreateDTOMock, userid);
+
+            result.IsCompleted.Should().BeTrue();
+            result.IsCompletedSuccessfully.Should().BeTrue();
+
+            await Task.CompletedTask;
```
Comment (YevheniiVovk) [inline Provis.UnitTests/Core/Services/WorkspaceServiceTests.cs]: Whe you don't use await in async invocations

File: Provis.UnitTests/Core/Services/WorkspaceServiceTests.cs [hunk 9]
```diff
@@ -87,5 +104,522 @@ public void TearDown()
...
+
+            SetupUpdateWorkspaceAsync();
+            SetupWorkspaceSaveChangesAsync();
+
+            var result = _workspaceService.UpdateWorkspaceAsync(workspaceCreateDTOMock);
+
+            result.IsCompleted.Should().BeTrue();
+            result.IsCompletedSuccessfully.Should().BeTrue();
+
+            await Task.CompletedTask;
```
Comment (YevheniiVovk) [inline Provis.UnitTests/Core/Services/WorkspaceServiceTests.cs:156]: same

File: Provis.UnitTests/Core/Services/WorkspaceServiceTests.cs [hunk 10]
```diff
@@ -87,5 +104,522 @@ public void TearDown()
...
+                .Where(x => x.StatusCode == HttpStatusCode.BadRequest)
+                .WithMessage(ErrorMessages.SendInviteYourself);
+        }
+
+        [Test]
+        [TestCase("2")]
+        public async Task SendInviteAsync_UserAcceptedInvite_ThrowHTTPException(string ownerId)
+        {
+            var workspaceInviteDTOMock = WorkspaceTestData.GetTestWorkspaceInviteDTO();
+            var userOwnerMock = UserTestData.GetTestUserList()[1];
```
Comment (YevheniiVovk) [inline Provis.UnitTests/Core/Services/WorkspaceServiceTests.cs]: Move it to with methods and fogot about static

File: Provis.UnitTests/Core/Services/WorkspaceServiceTests.cs [hunk 11]
```diff
@@ -87,5 +104,522 @@ public void TearDown()
...
+                .SendInviteAsync(workspaceInviteDTOMock, ownerId);
+
+            await act.Should()
+                .ThrowAsync<HttpException>()
+                .Where(x => x.StatusCode == HttpStatusCode.BadRequest)
+                .WithMessage(ErrorMessages.UserAcceptedInvite);
+        }
+
+        [Test]
+        [TestCase("1")]
```
Comment (YevheniiVovk) [inline Provis.UnitTests/Core/Services/WorkspaceServiceTests.cs]: For a what reason testcase here ?

File: Provis.UnitTests/Core/Services/WorkspaceServiceTests.cs [hunk 12]
```diff
@@ -87,5 +104,522 @@ public void TearDown()
...
+            SetupGetFirstBySpecAsync(userNull);
+
+            Func<Task> act = () => _workspaceService
+                .SendInviteAsync(workspaceInviteDTOMock, ownerId);
+
+            await act.Should()
+                .ThrowAsync<HttpException>()
+                .Where(x => x.StatusCode == HttpStatusCode.BadRequest)
+                .WithMessage(ErrorMessages.UserAlreadyHasInvite);
+        }
```
Comment (YevheniiVovk) [inline Provis.UnitTests/Core/Services/WorkspaceServiceTests.cs]: Same about static and async

File: Provis.UnitTests/Core/Services/WorkspaceServiceTests.cs [hunk 13]
```diff
@@ -87,5 +104,522 @@ public void TearDown()
...
+
+            SetupGetTemplateHtmlAsStringAsync(viewName, templateStringMock);
+            SetupSendEmailAsync();
+
+            var result = _workspaceService.SendInviteAsync(workspaceInviteDTOMock, ownerId);
+
+            result.IsCompleted.Should().BeTrue();
+            result.IsCompletedSuccessfully.Should().BeTrue();
+
+            await Task.CompletedTask;
```
Comment (YevheniiVovk) [inline Provis.UnitTests/Core/Services/WorkspaceServiceTests.cs]: Same about static and async

File: Provis.UnitTests/Core/Services/WorkspaceServiceTests.cs [hunk 14]
```diff
@@ -87,5 +104,522 @@ public void TearDown()
...
+        }
+
+        [Test]
+        [TestCase(1,"2")]
```
Comment (YevheniiVovk) [inline Provis.UnitTests/Core/Services/WorkspaceServiceTests.cs]: Please use testcasesourse

File: Provis.UnitTests/Core/Services/WorkspaceServiceTests.cs [hunk 15]
```diff
@@ -87,5 +104,522 @@ public void TearDown()
...
+        {
+            var inviteUserMock = InviteUserTestData.GetInviteUserList()[0];
+            SetupGetInviteUserByKeyAsync(inviteId, inviteUserMock);
+
+            var result = _workspaceService.DenyInviteAsync(inviteId, ownerId);
+
+            result.IsCompleted.Should().BeTrue();
+            result.IsCompletedSuccessfully.Should().BeTrue();
+
+            await Task.CompletedTask;
```
Comment (YevheniiVovk) [inline Provis.UnitTests/Core/Services/WorkspaceServiceTests.cs]: Same about static and async

File: Provis.UnitTests/Core/Services/WorkspaceServiceTests.cs [hunk 16]
```diff
@@ -87,5 +104,522 @@ public void TearDown()
...
+            var inviteUserMock = InviteUserTestData.GetInviteUserList()[0];
+            SetupGetInviteUserByKeyAsync(inviteId, inviteUserMock);
+
+            Func<Task> act = () => _workspaceService
+                .DenyInviteAsync(inviteId, ownerId);
+
+            await act.Should()
+                .ThrowAsync<HttpException>()
+                .Where(x => x.StatusCode == HttpStatusCode.BadRequest)
+                .WithMessage(ErrorMessages.CannotDenyInvite);
```
Comment (YevheniiVovk) [inline Provis.UnitTests/Core/Services/WorkspaceServiceTests.cs]: Same about static

File: Provis.UnitTests/Core/Services/WorkspaceServiceTests.cs [hunk 17]
```diff
@@ -87,5 +104,522 @@ public void TearDown()
...
+
+            await Task.CompletedTask;
+        }
+
+        [Test]
+        [TestCase("1")]
+        public async Task GetWorkspaceListAsync_UserIsValid_ReturnWorkspaceList(string userId)
+        {
+            var expectedWorkspaceListMock = WorkspaceTestData.GetTestWorkspaceListInfoDTO();
+            var userWorkspaceMock = UserWorkspaceTestData.GetTestUserWorkspaceList();
```
Comment (YevheniiVovk) [inline Provis.UnitTests/Core/Services/WorkspaceServiceTests.cs]: Same about static

File: Provis.UnitTests/Core/Services/WorkspaceServiceTests.cs [hunk 18]
```diff
@@ -87,5 +104,522 @@ public void TearDown()
...
+            var result = _workspaceService.DenyInviteAsync(inviteId, ownerId);
+
+            result.IsCompleted.Should().BeTrue();
+            result.IsCompletedSuccessfully.Should().BeTrue();
+
+            await Task.CompletedTask;
+        }
+
+        [Test]
+        [TestCase("1")]
```
Comment (YevheniiVovk) [inline Provis.UnitTests/Core/Services/WorkspaceServiceTests.cs]: Same about testcase

File: Provis.UnitTests/Core/Services/WorkspaceServiceTests.cs [hunk 19]
```diff
@@ -87,5 +104,522 @@ public void TearDown()
...
+
+            var result = await _workspaceService.GetWorkspaceListAsync(userId);
+
+            result.Should().NotBeNull();
+            result.Should().Equal(expectedWorkspaceListMock);
+        }
+
+        [Test]
+        [TestCase(1, "2")]
+        public async Task AcceptInviteAsync_CannotAcceptInvite_ThrowHTTPException(int inviteId, string userId)
```
Comment (YevheniiVovk) [inline Provis.UnitTests/Core/Services/WorkspaceServiceTests.cs]: Naming

File: Provis.UnitTests/Core/Services/WorkspaceServiceTests.cs [hunk 20]
```diff
@@ -87,5 +104,522 @@ public void TearDown()
...
+            _mapperMock.SetupMap(userWorkspaceMock, expectedWorkspaceListMock);
+
+            var result = await _workspaceService.GetWorkspaceListAsync(userId);
+
+            result.Should().NotBeNull();
+            result.Should().Equal(expectedWorkspaceListMock);
+        }
+
+        [Test]
+        [TestCase(1, "2")]
```
Comment (YevheniiVovk) [inline Provis.UnitTests/Core/Services/WorkspaceServiceTests.cs]: TestCaseSource

File: Provis.UnitTests/Core/Services/WorkspaceServiceTests.cs [hunk 21]
```diff
@@ -87,5 +104,522 @@ public void TearDown()
...
+            var userWorkspaceMock = UserWorkspaceTestData.GetTestUserWorkspaceList()[0];
+
+            SetupGetFirstBySpecAsync(userWorkspaceMock);
+            _mapperMock.SetupMap(userWorkspaceMock, expectedWorkspaceMock);
+
+            var result = await _workspaceService.GetWorkspaceInfoAsync(workspaceId, userId);
+
+            result.Should().NotBeNull();
+            result.Should().Be(expectedWorkspaceMock);
+        }
```
Comment (YevheniiVovk) [inline Provis.UnitTests/Core/Services/WorkspaceServiceTests.cs]: Same as previos

File: Provis.UnitTests/Core/Services/WorkspaceServiceTests.cs [hunk 22]
```diff
@@ -87,5 +104,522 @@ public void TearDown()
...
+
+        [Test]
+        [TestCase("2")]
```
Comment (YevheniiVovk) [inline Provis.UnitTests/Core/Services/WorkspaceServiceTests.cs]: For a what reason TestCase ?

File: Provis.UnitTests/Core/Services/WorkspaceServiceTests.cs [hunk 23]
```diff
@@ -87,5 +104,522 @@ public void TearDown()
...
+                                                           It.IsAny<WorkspaceInvite>()))
+                .ReturnsAsync(templateInstance)
+                .Verifiable();
+        }
+
+        protected void SetupSendEmailAsync()
+        {
+            _emailSendServiceMock
+                .Setup(x => x.SendEmailAsync(It.IsAny<MailRequest>()))
+                .Returns(Task.CompletedTask)
```
Comment (YevheniiVovk) [inline Provis.UnitTests/Core/Services/WorkspaceServiceTests.cs]: Are any better way to return async Task

File: Provis.UnitTests/Core/Services/WorkspaceServiceTests.cs [hunk 24]
```diff
@@ -87,5 +104,522 @@ public void TearDown()
...
+                .Setup(x => x.GetListBySpecAsync(It.IsAny<ISpecification<UserWorkspace>>()))
+                .ReturnsAsync(listInstance)
+                .Verifiable();
+        }
+
+        protected void SetupGetFirstBySpecAsync(UserWorkspace userWorkspaceInstance)
+        {
+            _userWorkspaceRepositoryMock
+                .Setup(x => x.GetFirstBySpecAsync(It.IsAny<ISpecification<UserWorkspace>>()))
+                .Returns(Task.FromResult(userWorkspaceInstance))
```
Comment (YevheniiVovk) [inline Provis.UnitTests/Core/Services/WorkspaceServiceTests.cs]: Let's use returnAsync method

File: Provis.UnitTests/Core/Services/WorkspaceServiceTests.cs [hunk 25]
```diff
@@ -87,5 +104,522 @@ public void TearDown()
...
+                .Setup(x => x.GetByKeyAsync(workspaceId))
+                .Returns(Task.FromResult(workspaceInstance))
+                .Verifiable();
+        }
+
+        protected void SetupGetInviteUserByKeyAsync(int inviteId, InviteUser inviteUserInstance)
+        {
+            _inviteUserRepositoryMock
+                .Setup(x => x.GetByKeyAsync(inviteId))
+                .Returns(Task.FromResult(inviteUserInstance))
```
Comment (YevheniiVovk) [inline Provis.UnitTests/Core/Services/WorkspaceServiceTests.cs]: ReturnAsync

File: Provis.UnitTests/Core/Services/WorkspaceServiceTests.cs [hunk 26]
```diff
@@ -87,5 +104,522 @@ public void TearDown()
...
+                .Setup(x => x.Map(source ?? It.IsAny<TSource>(), destination))
+                .Returns(destination)
+                .Verifiable();
+        }
+
+        protected void SetupGetWorkspaceByKeyAsync(int workspaceId, Workspace workspaceInstance)
+        {
+            _workspaceRepositoryMock
+                .Setup(x => x.GetByKeyAsync(workspaceId))
+                .Returns(Task.FromResult(workspaceInstance))
```
Comment (YevheniiVovk) [inline Provis.UnitTests/Core/Services/WorkspaceServiceTests.cs]: ReturnAsync

File: Provis.UnitTests/Core/Services/WorkspaceServiceTests.cs [hunk 27]
```diff
@@ -87,5 +104,522 @@ public void TearDown()
...
+
+                    {
+                        WorkSpaceRoles.ManagerId, new List<WorkSpaceRoles>
+                        {
+                            WorkSpaceRoles.ManagerId, WorkSpaceRoles.MemberId, WorkSpaceRoles.ViewerId
+                        } 
+                    }
+                }
+            };
+        }
```
Comment (YevheniiVovk) [inline Provis.UnitTests/Core/Services/WorkspaceServiceTests.cs:646]: Remove static

**Review by FeloniukM — APPROVED** [review]

**Pull request merged**